src/igmpproxy.c

```diff
@@ -140,20 +140,36 @@ int main( int ArgCn, char *ArgVc[] ) {
             break;
         }
 
-        if ( !NotAsDaemon ) {
+        // Chroot and drop privileges.
+        {
+            struct Config *config;
+            struct passwd *pw;
 
-            // Only daemon goes past this line...
-            if (fork()) exit(0);
+            config = getCommonConfig();
 
-            // Detach daemon from terminal
-            if ( close( 0 ) < 0 || close( 1 ) < 0 || close( 2 ) < 0
-                || open( "/dev/null", 0 ) != 0 || dup2( 0, 1 ) < 0 || dup2( 0, 2 ) < 0
-                || setpgid( 0, 0 ) < 0
-            ) {
-                my_log( LOG_ERR, errno, "failed to detach daemon" );
+            if (strlen(config->user) != 0) {
+                pw = getpwnam(config->user);
+                if (pw == NULL)
+                    my_log(LOG_ERR, 0, "unknown user %s", config->user);
             }
+
+            if (strlen(config->chroot) != 0)
+                if (chroot(config->chroot) != 0 || chdir("/") != 0)
+                    my_log(LOG_ERR, 0, "unable to chroot to %s",
+                      config->chroot);
+
+            if (strlen(config->user) != 0)
+                if (setgroups(1, &pw->pw_gid) != 0 ||
+                  setresgid(pw->pw_gid, pw->pw_gid, pw->pw_gid) != 0 ||
+                  setresuid(pw->pw_uid, pw->pw_uid, pw->pw_uid) != 0)
+                    my_log(LOG_ERR, 0, "unable to drop privileges to %s",
+                      config->user);
         }
 
+        if (!NotAsDaemon)
+            if (daemon(1, 0) == -1)
```

Comment:
Let me explain it again: After executing chroot and chdir you have no access to original root filesystem. So trying to open /dev/null or any other file from original file system either directly via syscall `open` or indirectly via `daemon` function just fails.

So you cannot use use neither `open` nor `daemon` function after calling `chroot` && `chdir`.

Reply:
You have to open all required files, including `/dev/null` prior executing `chroot` syscall.

Reply:
/dev/null is now opened before chrooting. Run tested on Alpine (musl libc), Arch (glibc) and OpenBSD.